Implement EPD Controller for newer Onyx Boox devices (Qualcomm)

Galunid opened the pull request:
Currently onyx devices rely on system controller, which doesn't work as well as it could with KOReader.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/koreader/android-luajit-launcher/250)
<!-- Reviewable:end -->

Galunid commented:
Any idea why it doesn't refresh screen when `OnyxEPDController.setEpdMode` is not called on lua side? Do I need to register it somewhere else, other than in `EDPFactory` and `DeviceInfo`? I get log that it uses `OnyxEpdController`, but it doesn't do anything. It works in `EPDTestActivity`, like it's supposed to.

pazos commented on app/build.gradle:
```diff
@@ -50,6 +50,7 @@ android {
         }
         arm {
             ndk { abiFilters "armeabi-v7a" }
+            minSdkVersion safSdk
```
I'm afraid we cannot merge this directly. That would mean that the minimum android version is now Lollipop and that's a big step since most eink readers are still 4.x.

Galunid replied:
Yup, it's a hack to get things testing quickly, it needs minSdkVersion >= 16, I'll probably add a new target for onyx. Same thing with modification in `EPDTestActivity`. I'll revert those when I'm preparing to merge.

pazos commented on app/build.gradle:
```diff
@@ -112,6 +113,9 @@ dependencies {
     implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_plugin_version"
 
+    // OnyxSDK
+    implementation('com.onyx.android.sdk:onyxsdk-base:1.5.4')
```
You can try to implement what onyx is doing under the hood, since it will be based on reflection, like us are doing.

pazos replied:
For example, there's a method called [`refreshScreenRegion`](https://github.com/buggins/coolreader/blob/master/android/src/com/onyx/android/sdk/api/device/epd/EpdController.java#L106-L108) which just issues a screen update for the current device. Assuming the device is IMX7 is just does https://github.com/buggins/coolreader/blob/master/android/src/com/onyx/android/sdk/device/IMX7Device.java#L151

Transforming that to a driver that skips the need of onyx sdk is not hard, but needs some time for trial and error. See https://github.com/koreader/android-luajit-launcher/blob/master/app/src/org/koreader/launcher/device/epd/freescale/NTXEPDController.kt#L41-L58.

pazos commented on app/manifest/base.xml:
```diff
@@ -16,7 +17,7 @@
     <uses-permission android:name="android.permission.WRITE_SETTINGS" />
 
-    <application
+    <application tools:replace="android:allowBackup"
```
Is that needed?, why?

Galunid replied:
It's needed since Onyxsdk's manifest has `android:allowBackup = true`. This introduces conflict in our manifest and gradle screams at me for doing so and refuese to build ;-).

pazos commented:
> Any idea why it doesn't refresh screen when `OnyxEPDController.setEpdMode` is not called on lua side? Do I need to register it somewhere else, other than in `EDPFactory` and `DeviceInfo`? I get log that it uses `OnyxEpdController`, but it doesn't do anything. It works in `EPDTestActivity`, like it's supposed to.

EPDTestActivity is hooked on android view hierarchy. The native window is not. That's why we pass a view to `setEpd` things. 

I would suggest to look at `/proc/cpuinfo` and see if it has a similar SoC of one of these supported. I think it should work fine with Nook EPD Controller, if you want to perform just full screen refreshes with it.

pazos commented:
> Do I need to register it somewhere else, other than in `EDPFactory` and `DeviceInfo`? 

Also yeah. Give it a look at https://github.com/koreader/android-luajit-launcher/blob/master/app/src/org/koreader/launcher/device/DeviceInfo.kt#L225-L250

Galunid commented:
I want to have partial refreshes, I added full refresh, to make certain it's handled by my controller, instead of Onyx's

Galunid commented:
> I would suggest to look at `/proc/cpuinfo` and see if it has a similar SoC of one of these supported. I think it should work fine with Nook EPD Controller, if you want to perform just full screen refreshes with it.

SoC is Qualcomm MSM8953 (Snapdragon 625)
Nook's EPD Controller doesn't work here. It doesn't get any exception, but it doesn't refresh screen either. Onyx has different modes than `freescale`/`rockchip`.

Galunid commented:
Notes to self and anyone interested
Example refresh log by normal usage of the device:
`I SDM     : update_to_display[0/1] -- marker[2903] waveform_mode = 4, update_mode = 0, Rect[0 0 1872 1404], flags = 8000`
From what I've found there are 4 `waveform_mode` values that user can choose:
`Normal mode` - 255
`Regal mode` - 5
`A2 mode` - 1
`X mode` - 4
additionally there is a full update with mode 2
`update_mode` is either 1 or 0, I haven't encountered any other, but that doesn't mean there isn't. 1 is for full refresh, 0 is for other types I believe.
There's enum in OnyxSdk, but it's used internally by device's implementations, to convert to `UpdateMode`
```java
public enum EPDMode {
    FULL,
    AUTO,
    TEXT,
    AUTO_PART,
    AUTO_BLACK_WHITE,
    AUTO_A2,
    EPD_REGLA
}
```

```java
public enum UpdateMode {
    None,
    DU,
    GU,
    GU_FAST,
    GC,
    ANIMATION,
    ANIMATION_QUALITY,
    ANIMATION_MONO,
    ANIMATION_X,
    GC4,
    REGAL,
    REGAL_D,
    DU_QUALITY,
    HAND_WRITING_REPAINT_MODE
}
```
Other interesting things
```java
int staticIntFieldSafely = ReflectUtil.getStaticIntFieldSafely(classForName, "EINK_ONYX_AUTO_MASK");
        int staticIntFieldSafely2 = ReflectUtil.getStaticIntFieldSafely(classForName, "EINK_ONYX_GC_MASK");
        int staticIntFieldSafely3 = ReflectUtil.getStaticIntFieldSafely(classForName, "EINK_AUTO_MODE_REGIONAL");
        int staticIntFieldSafely4 = ReflectUtil.getStaticIntFieldSafely(classForName, "EINK_WAIT_MODE_NOWAIT");
        int staticIntFieldSafely5 = ReflectUtil.getStaticIntFieldSafely(classForName, "EINK_WAIT_MODE_WAIT");
        int staticIntFieldSafely6 = ReflectUtil.getStaticIntFieldSafely(classForName, "EINK_WAVEFORM_MODE_DU");
        int staticIntFieldSafely7 = ReflectUtil.getStaticIntFieldSafely(classForName, "EINK_WAVEFORM_MODE_ANIM");
        int staticIntFieldSafely8 = ReflectUtil.getStaticIntFieldSafely(classForName, "EINK_WAVEFORM_MODE_GC4");
        int staticIntFieldSafely9 = ReflectUtil.getStaticIntFieldSafely(classForName, "EINK_WAVEFORM_MODE_GC16");
        int staticIntFieldSafely10 = ReflectUtil.getStaticIntFieldSafely(classForName, "EINK_WAVEFORM_MODE_REAGL");
        int staticIntFieldSafely11 = ReflectUtil.getStaticIntFieldSafely(classForName, "EINK_REAGL_MODE_REAGLD");
        int staticIntFieldSafely12 = ReflectUtil.getStaticIntFieldSafely(classForName, "EINK_UPDATE_MODE_PARTIAL");
        int staticIntFieldSafely13 = ReflectUtil.getStaticIntFieldSafely(classForName, "EINK_UPDATE_MODE_FULL");
        int staticIntFieldSafely14 = ReflectUtil.getStaticIntFieldSafely(classForName, "UI_DEFAULT_MODE");
        ReflectUtil.getStaticIntFieldSafely(classForName, "EINK_DITHER_MODE_DITHER");
        d = staticIntFieldSafely;
        e = staticIntFieldSafely2;
        f = staticIntFieldSafely3 | staticIntFieldSafely4 | staticIntFieldSafely6 | staticIntFieldSafely12;
        g = ReflectUtil.getStaticIntFieldSafely(classForName, "UI_DU_QUALITY_MODE");
        h = staticIntFieldSafely3 | staticIntFieldSafely4 | staticIntFieldSafely9 | staticIntFieldSafely12;
        i = staticIntFieldSafely3 | staticIntFieldSafely5 | staticIntFieldSafely9 | staticIntFieldSafely13;
        j = staticIntFieldSafely3 | staticIntFieldSafely4 | staticIntFieldSafely7 | staticIntFieldSafely12;
        k = ReflectUtil.getStaticIntFieldSafely(classForName, "UI_A2_QUALITY_MODE");
        l = ReflectUtil.getStaticIntFieldSafely(classForName, "UI_MONO_A2_MODE");
        m = ReflectUtil.getStaticIntFieldSafely(classForName, "UI_X_A2_MODE");
        n = staticIntFieldSafely3 | staticIntFieldSafely4 | staticIntFieldSafely8 | staticIntFieldSafely12;
        o = staticIntFieldSafely3 | staticIntFieldSafely4 | staticIntFieldSafely10 | staticIntFieldSafely12;
        p = staticIntFieldSafely3 | staticIntFieldSafely4 | staticIntFieldSafely11 | staticIntFieldSafely10 | staticIntFieldSafely12;
        q = staticIntFieldSafely14;
        r = ReflectUtil.getStaticIntFieldSafely(classForName, "HAND_WRITING_REPAINT_MODE");
        Class<?> classForName2 = ReflectUtil.classForName("android.onyx.optimization.Constant");
        // used by isInSystemRefreshModeDefault
        UPDATE_MODE_DEFAULT = ReflectUtil.getStaticIntFieldSafely(classForName2, "UPDATE_MODE_DEFAULT");
        // not used anywhere in code
        UPDATE_MODE_DU = ReflectUtil.getStaticIntFieldSafely(classForName2, "UPDATE_MODE_DU");
        // not used anywhere in code
        UPDATE_MODE_A2 = ReflectUtil.getStaticIntFieldSafely(classForName2, "UPDATE_MODE_A2");
        // not used anywhere in code
        UPDATE_MODE_REGAL = ReflectUtil.getStaticIntFieldSafely(classForName2, "UPDATE_MODE_REGAL");
        // used by isInSystemRefreshModeX
        UPDATE_MODE_X = ReflectUtil.getStaticIntFieldSafely(classForName2, "UPDATE_MODE_X");
```

pazos commented:
> SoC is Qualcomm MSM8953 (Snapdragon 625)

Good, new platform :)

> Nook's EPD Controller doesn't work here. It doesn't get any exception, but it doesn't refresh screen either. Onyx has different modes than freescale/rockchip.

Also good :). It is not expected to work as is. But all supported platforms follow the same scheme:

- there's a method in the view class that's in charge of requesting a screen update
- we invoke that method using reflection, that's useful as we just need to know the name of the class and the parameters it takes.
- if there's no sdk, which is usual, the stock app is re'd to know the name of the method and its parameters

In the case of Tolinos (and other full_eink devices) the system driver doesn't do any refresh on its own and we do them all.
In the rest of cases the system driver update the window on its own and we just flash on user request.

In the case of onyx, I think there's a way to prevent system updates (some sort of manual mode) and then act like in Tolinos. If you don't prevent system updates you'll get 2x screen refreshes.

Galunid commented:
It "works". The bad news is that system applications don't work, after going home, there's no screen refresh etc. And KOReader doesn't always update properly, but it's a start.

Galunid commented:
Sidenote, full screen refresh in KOReader on "normal" mode is a bit slower than "onyx A2" refresh, where A2 degrades quality quite a bit. The improvement is drastic.

Galunid commented:
When calling `refreshScreen`
```java
public enum UpdateMode {
    None, // 5
    DU, // 1
    GU, // 2
    GU_FAST, // 1
    GC, // 98
    ANIMATION, // 4
    ANIMATION_QUALITY, // 2308
    ANIMATION_MONO, // 33554436
    ANIMATION_X, // 16777220
    GC4, // 3
    REGAL, // 6
    REGAL_D, // 4102
    DU_QUALITY, // 2305
    HAND_WRITING_REPAINT_MODE // 524290
}
```

pazos commented on app/src/org/koreader/launcher/device/epd/onyx/OnyxEPDController.kt:
```diff
@@ -0,0 +1,72 @@
+        }
+        object : Thread(){
+            override fun run(){
+                EpdController.enableScreenUpdate(targetView, true)
```
I think we can add `onResume` and `onPause` methods to EPDInterface, enable manual mode on resume and disable it on pause.

This way you can avoid the thread, the delay and just fire `EpdController.refreshScreenRegion(targetView, x, y, x + width, y + height, updateMode)`.

BTW, the final waveform depends on https://github.com/koreader/koreader-base/blob/master/ffi/framebuffer_android.lua, which depends on proper DeviceInfo ids, like HAS_EINK_SUPPORT and HAS_EINK_FULL_SUPPORT

Galunid replied:
Well, this doesn't work anyway, the refresh is queued and after enabling refresh again, the queue gets emptied, regardless of delay. Ironically, `enableScreenUpdate` blocks our updates too... I asked in https://github.com/onyx-intl/OnyxAndroidDemo/issues/40 if there's a way to prevent updates, but I'm not too optimistic about getting a response ;)

Galunid replied:
For now, I'd like to get it working without onyx sdk, since there are multiple reasons for not using it, we need like 2 functions.

Galunid commented:
This "works". The refresh mode is always set to `94` (full refresh). Mapping other modes shouldn't be a problem.
Unfortunately it's wonky (**refresh happens before page is changed**).
@pazos Any idea why that may happen?

pazos commented on app/src/org/koreader/launcher/device/epd/onyx/OnyxEPDController.kt:
```diff
@@ -0,0 +1,41 @@
+import android.util.Log
+import android.view.View
+
+import com.onyx.android.sdk.api.device.epd.EpdController
```
these aren't needed now, right?

Galunid replied:
Yes, it was already removed in https://github.com/koreader/android-luajit-launcher/pull/250/commits/01ad467ba1daf6731ee700a3f8d5ced3680765a1

pazos commented:
> This "works". The refresh mode is always set to `98` (full refresh). Mapping other modes shouldn't be a problem.
> Unfortunately it's wonky (**refresh happens before page is changed**).
> @pazos Any idea why that may happen?

Emm, I see you are somehow using the same settings that the rest of freescale devices. Also you implemented a delay with a thread, that's sounds sensible. 

But, **again**, you need to tweak https://github.com/koreader/koreader-base/blob/master/ffi/framebuffer_android.lua to use the waveform modes and the delays you want.

For example, see how Freescale devices are now handling `refreshPartialImp(x, y, w, h)` https://github.com/koreader/koreader-base/blob/master/ffi/framebuffer_android.lua#L152. You need to add a delay other than 0 and write the specific waveform mode.

pazos commented on app/src/org/koreader/launcher/device/DeviceInfo.kt:
```diff
@@ -259,7 +261,7 @@ object DeviceInfo {
 
         // needs a view
-        NEEDS_VIEW = ! EINK_SUPPORT || EINK_FREESCALE
+        NEEDS_VIEW = ! EINK_SUPPORT || EINK_FREESCALE || EINK_QCOM
```
Not sure if it was legible enough.

NEEDS_VIEW is used in MainActivity to create a surfaceView and take control of the native window from the java side. That's used to perform epd updates, because passing the ViewRootImpl() as is doesn't work.

The logic right now is: enable the surfaceView **unless** it is a eink device and not a freescale one. My advice would be check first if the device works without the surfaceView as the surfaceView is costly and might introduce some delay.

Galunid replied:
I removed it and there's indeed quite a bit difference. It works way better now. Earlier I had issues with navigating menu I didn't know what was causing them. They're gone now. I guess it wasn't legible enough, I took it as "you want to get `view` as a parameter, then set this to true".

pazos replied:
Probably doesn't make any sense now to have it on DeviceInfo. The surfaceView hack only makes sense on Nooks and Tolinos (EINK_FREESCALE). It is also needed on androidTV and ChromeOS because CRe documents went black if there's no input for a few seconds.

Something like the following could make sense

```kotlin

/* Whether this device needs a SurfaceView implementation or not. Most devices work well (and faster) without one.
Only enable it if your device needs it for EPD updates (like Tolinos) or somehow the system driver work badly without one
(certain android TV boxes, which dim the screen if they don't detect a "valid" view hierarchy) */

NEEDS_VIEW = EINK_FREESCALE
```

pazos reviewed (changes requested):
Looks way better now, congrats!!

Just 2cts to look even better:

- move the core logic to an abstract class called `QualcommEPDController` and wrap reflection methods in a try catch block. 
- rename the class that implements the EPDInterface to something less generic, like `QualcommOnyxEPDController` and extend from `QualcommEPDController`

That way we can add calls to the abstract class directly in `EPDTestActivity` without crashes, and you can add all the comments you want in the abstract class.

Please have a look at the difference between `NTXEPDController` and `NTXNewEPDController` to understand what I'm talking about.

Galunid commented:
It's working pretty well, at least better than the default in my opinion, some things can definitely be improved with fine-tuning stuff in base. What's left to do is implement `onResume`/`onPause`, since we need to disable updates only once, not on every refresh. And setting default mode to A2, and disabling "system" full refreshes, since those are now handled by KOReader. Other than that it needs a bit of refactoring and safeguards against reflection crashes

Galunid commented:
Just to confirm, adding `epd.onPause()` in `MainActivity.kt` shouldn't crash anything, if I created default method in interface with empty implementation?
```diff
@@ -126,6 +126,7 @@ class MainActivity : BaseActivity() {
     override fun onResume() {
         Logger.d(TAG_MAIN, "onResume()")
         super.onResume()
+        epd.onResume()
         applyCustomTimeout(true)
     }
 
@@ -133,6 +134,7 @@ class MainActivity : BaseActivity() {
     override fun onPause() {
         Logger.d(TAG_MAIN, "onPause()")
         super.onPause()
+        epd.onPause()

interface EPDInterface {
+    fun onPause(){}
+    fun onResume(){}
```

pazos commented:
@Galunid: an interface is like an abstract class but with all methods unimplemented. Classes that implement an interface need to be abstract or implement all methods from the interface.

So, if you add to EPDInterface

```kotlin
fun onPause()
fun onResume()
```

you need to implement these methods in all classes that implement the interface. Everywhere, except on onyx, the implementation is empty:

```kotlin
override fun onPause() {}
override fun onResume() {}
```

Pretty much like we're doing with [some lights methods](https://github.com/koreader/android-luajit-launcher/blob/master/app/src/org/koreader/launcher/device/lights/OnyxWarmthController.kt#L103-L105), but this time is simpler as both onResume and onPause are void methods

pazos commented:
^^ also applies, somehow, to the logic I requested.

For example:

1. [`NTXEPDController`]() is an abstract class, with a single static method `requestEpdMode`, which performs, via reflection, an EPD update. It is cool because we can repurpose the abstract class as is everywhere and will run **without crashes** even on incompatible devices (ie:  no unhandled exceptions if reflection fails)

2. [`NTXEPDNewController`](https://github.com/koreader/android-luajit-launcher/blob/master/app/src/org/koreader/launcher/device/epd/freescale/NTXNewEPDController.kt) is used as a bridge that implements the only method of [`EPDInterface`](https://github.com/koreader/android-luajit-launcher/blob/master/app/src/org/koreader/launcher/interfaces/EPDInterface.kt) inheriting the implementation from NTXEPDControler

So, this scheme applies to all supported devices. 1 is the implementation, as an abstract class with static methods. 2 is the bridge.

If you add methods to EPDInterface you'll need to implement them on every bridge (ie: all device/epd classes that are not abstract) but you don't need to touch their implementations.

I hope it is clear enough :)

Galunid commented:
Yes that's pretty reasonable, what I meant in my earlier question was `default method`. It was introduced in `Java 8` I believe.
```java
interface LightsManager{
    default void toggleLights(){
        System.out.println("Toggling lights, yayy!");
    }
}
class TestLightsManager implements LightsManager{

}

public class Main{
    public static void main(String[] args){
        LightsManager lm = new TestLightsManager();
        lm.toggleLights(); // this works
    }
}
```

This way you can add methods to the interfaces without breaking backwards compatibility (which perfectly fit this case), not sure how that works in Kotlin though. Anyway that's not needed, since I need to call `preventSystemRefresh` before our every refresh ;/

P.S. Really appreciate the help!

Galunid commented:
Sadly `NEEDS_VIEW` is needed, otherwise system "overrides" our update.
Note to self: request A2 mode in `preventSystemUpdate`
Tomorrow I'll open pull request in base with the rest of the changes, I'm think I'm done with this part ;)

pazos commented:
I tried to submit a PR to your branch but it is too old :(. Could you add that? (please review the **mode** passed. I took 98 from old conversation, it might be wrong.

I've tested and saw nothing breaks in EPDTestActivity, so it is good to go from my POV but I would like to wait for the base counterpart before merging this one.

```diff
diff --git a/app/res/layout/main.xml b/app/res/layout/main.xml
index e636053..c9d1024 100644
--- a/app/res/layout/main.xml
+++ b/app/res/layout/main.xml
@@ -75,6 +75,24 @@
             android:text="freescale/ntx update"
             android:textSize="28sp" />
 
+        <TextView
+            android:layout_width="fill_parent"
+            android:layout_height="12dp" />
+
+        <TextView
+            android:id="@+id/qualcommText"
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            android:gravity="center_horizontal"
+            android:textSize="18sp" />
+
+        <Button
+            android:id="@+id/qualcommButton"
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            android:text="qualcomm/onyx update"
+            android:textSize="28sp" />
+
         <TextView
             android:layout_width="fill_parent"
             android:layout_height="18dp" />
diff --git a/app/src/org/koreader/launcher/EPDTestActivity.kt b/app/src/org/koreader/launcher/EPDTestActivity.kt
index 6bf345c..2059a38 100644
--- a/app/src/org/koreader/launcher/EPDTestActivity.kt
+++ b/app/src/org/koreader/launcher/EPDTestActivity.kt
@@ -10,6 +10,7 @@ import android.widget.TextView
 import android.widget.Button
 
 import org.koreader.launcher.device.epd.freescale.NTXEPDController
+import org.koreader.launcher.device.epd.qualcomm.QualcommEPDController
 import org.koreader.launcher.device.epd.rockchip.RK30xxEPDController
 import org.koreader.launcher.device.epd.rockchip.RK33xxEPDController
 
@@ -28,9 +29,11 @@ class EPDTestActivity : Activity() {
         val rk30xxDescription: TextView = findViewById(R.id.rk30xxText)
         val rk33xxDescription: TextView = findViewById(R.id.rk33xxText)
         val ntxNewDescription: TextView = findViewById(R.id.ntxNewText)
+        val qualcommDescription: TextView = findViewById(R.id.qualcommText)
         val rk30xxButton: Button = findViewById(R.id.rk30xxButton)
         val rk33xxButton: Button = findViewById(R.id.rk33xxButton)
         val ntxNewButton: Button = findViewById(R.id.ntxNewButton)
+        val qualcommButton: Button = findViewById(R.id.qualcommButton)
         val shareButton: Button = findViewById(R.id.shareButton)
 
         /* current device info */
@@ -68,6 +71,10 @@ class EPDTestActivity : Activity() {
         ntxNewDescription.text = "This button should work on modern Tolinos/Nooks and other ntx boards"
         ntxNewButton.setOnClickListener { runEinkTest(NTXTEST) }
 
+        /* qualcomm - At least Onyx Boox Nova 2 */
+        qualcommDescription.text = "This button should work on some Onyx Boox devices. At least qualcomm ones"
+        qualcommButton.setOnClickListener { runEinkTest(QUALCOMMTEST) }
+
         /* share button */
         shareButton.setOnClickListener { shareText(info.text.toString()) }
     }
@@ -84,15 +91,20 @@ class EPDTestActivity : Activity() {
             } else if (test == RK33xxTEST) {
                 info.append("rk33xx: ")
                 if (RK33xxEPDController.requestEpdMode("EPD_FULL")) success = true
-            } else if (test == NTXTEST) {
+            } else if (test == NTXTEST || test == QUALCOMMTEST) {
                 // get screen width and height
                 val display = windowManager.defaultDisplay
                 val size = Point()
                 display.getSize(size)
                 val width = size.x
                 val height = size.y
-                info.append("tolino: ")
-                if (NTXEPDController.requestEpdMode(v, 34, 50, 0, 0, width, height)) success = true
+                if (test == NTXTEST) {
+                    info.append("tolino: ")
+                    if (NTXEPDController.requestEpdMode(v, 34, 50, 0, 0, width, height)) success = true
+                } else if (test == QUALCOMMTEST) {
+                    info.append("qualcomm: ")
+                    if (QualcommEPDController.requestEpdMode(v, 98, 50, 0, 0, width, height)) success = true
+                }
             }
         } catch (e: Exception) {
             e.printStackTrace()
@@ -115,6 +127,7 @@ class EPDTestActivity : Activity() {
         private const val RK30xxTEST = 1
         private const val RK33xxTEST = 2
         private const val NTXTEST = 3
+        private const val QUALCOMMTEST = 4
         private val MANUFACTURER = android.os.Build.MANUFACTURER
         private val BRAND = android.os.Build.BRAND
         private val MODEL = android.os.Build.MODEL
```

Galunid commented:
@pazos Your patch works as expected, I rebased and squashed the commits, in case you need to push anything. It's still missing one thing and doesn't work as well as the default yet (sometimes menu gets stuck), so I'd prefer to hold it for a while too, not to mention that without base it'll break, since it'll get `freescale` refresh modes.

pazos commented:
> Since we are dealing with View I had a vague recollection that views should be updated on a separate UI thread, does that apply here (should I use something like runOnUIThread?)

Nope. The view is just a hint for the framework that tries to perform an update on the root view of the activity. On rockchip devices we use [the native window directly as the rootView](https://github.com/koreader/android-luajit-launcher/blob/master/app/src/org/koreader/launcher/MainActivity.kt#L426) and on freescale we hook a surfaceView behind the native window and pass that instead.

Unless you find a crash saying "method x tried to touch the view hierarchy (...) only from UIThread" or "Looper ..." , in that case you'll need to pass the activity instead of a view and run with `activity.runOnUiThread`

pazos reviewed (approved):
LGTM :+1:

Galunid commented:
Other than that I need to force the device into A2 mode, to prevent updates without user interaction, please hold it for a while.

pazos commented:
> Other than that I need to force the device into A2 mode, to prevent updates without user interaction, please hold it for a while.

I plan to merge https://github.com/koreader/android-luajit-launcher/pull/257 later this week . I added the resume and pause methods to DeviceHelper, which are propagated to EPD controllers.

Please consider to work on top of that branch if you want to disable/enable things onResume/onPause. :+1:

Galunid commented:
@pazos This PR is ready to be merged after the one in base, users will have to select the mode manually I guess, since there's no api that works within onyxsdk.

pazos commented:
@Galunid: if you don't mind waiting a couple of days I would like to merge #257 first and then resolve conflicts on this one.

I see that in the end you won't need to run methods onResume() and onPause(). I guess I will revert that bit from #257 unless you want to experiment with them on the future.

Galunid commented:
@pazos I think it's fine to keep them, I don't think there are any conflicts here actually, I'll just have to add onPause/onResume, which isn't a big deal. This can wait as long as you need.

Galunid commented:
Rebased, resolved conflicts, untested yet, but it shouldn't break, unless I messed up resolving conflicts ^_^
I'll report once I test it.

pazos commented:
@Galunid: everything looks fine. Going to merge this one and submit https://github.com/koreader/koreader-base/pull/1221#issuecomment-717218405 on top of it.

If you find a bug or something you want to correct we still have time, since we're not bumping this right now :)

mergen3107 commented:
@Galunid 
Sorry I lost track of this in the middle of this thread, so Onyx Boox devices with Qualcomm SoC will only work in A2 mode in KOReader? Do you know whether this degrades image quality in muPDF severely compared to Normal mode?

Galunid commented:
@mergen3107 Nope, they will work in normal mode, but require the "A2" mode enabled in onyx application settings. Otherwise you'll get extra flashes. So no, you'll effectively have everything rendered in normal mode

mergen3107 commented:
@Galunid 
Ah, thanks! :)

mergen3107 commented:
@Galunid 
Hi, sorry for bumping this old thread with a somewhat off topic question, but do you know by chance how to full refresh screen on an Onyx Boox via command line?

mergen3107 commented:
I found the answer!

`adb shell am broadcast -a onyx.android.intent.action.REFRESH_SCREEN`
